File: data_structures/arrays/004Sort012/zeronetwo.cpp
#include <bits/stdc++.h>
using namespace std;

// O(n) time O(1) space but not optimal
void count_zero_one_two(int array[], int len) {
  int zero = 0;
  int one = 0;
  int two = 0;
  for (int i = 0; i < len; i++) {
    if (array[i] == 0) {
      zero++;
    }
    if (array[i] == 1) {
      one++;
    }
    if (array[i] == 2) {
      two++;
    }
  }
  int i = 0;
  while (zero--) {
    array[i++] = 0;
  }
  while (one--) {
    array[i++] = 1;
  }
  while (two--) {
    array[i++] = 2;
  }
}

void dutch_national_flag(int array[], int len) {
  int low = 0;
  int high = len - 1;
  int mid = 0;

  while (mid <= high) {
    switch (array[mid]) {
      case 0:
        swap(array[low++], array[mid++]);
        break;
      case 1:
        mid++;
        break;
      case 2:
        swap(array[high--], array[mid]);
    }
  }
}

int main() {
  // codes
  int array[] = {1, 0, 2, 1, 0, 0, 0, 1, 2, 2, 2, 1, 0};
  int len = sizeof(array) / sizeof(array[0]);
  // count_zero_one_two(array, len);
  dutch_national_flag(array, len);
  for (int i = 0; i < len; i++) {
    cout << array[i] << " ";
  }
  cout << "\n";
}

File: data_structures/arrays/004Sort012/zeronetwo.py
def dutch_national_flag(array: list, first_element: int, middle_element: int, final_element: int) -> None:
    low = 0
    high = len(array) - 1
    mid = 0

    while mid <= high:
        if array[mid] == first_element:
            array[mid], array[low] = array[low], array[mid]
            mid += 1
            low += 1
            continue
        if array[mid] == middle_element:
            mid += 1
            continue
        if array[mid] == final_element:
            array[mid], array[high] = array[high], array[mid]
            high -= 1
            continue


def main():
    array = [0,1,2,1,2,0,1,0,1,2,1,1,1,0,0,2]
    dutch_national_flag(array, first_element=0, middle_element=1,final_element=2)
    print("Array Is Now Sorted!")
    print(array)


if __name__ == "__main__":
    main()